lib/route/nh.c

```diff
@@ -0,0 +1,529 @@
+/* SPDX-License-Identnhier: LGPL-2.1-only */
+/*
+ * Copyright (c) 2022 Stanislav Zaikin <stanislav.zaikin@46labs.com>
+ */
+
+#include <linux-private/linux/nexthop.h>
+#include <netlink-private/netlink.h>
+#include <netlink/route/nh.h>
+#include <netlink/hashtable.h>
+
+/** @cond SKIP */
+#define NH_ATTR_FLAGS			(1 << 0)
+#define NH_ATTR_ID			(1 << 1)
+#define NH_ATTR_GATEWAY			(1 << 2)
+#define NH_ATTR_OIF			(1 << 3)
+#define NH_ATTR_GROUP			(1 << 4)
+#define NH_ATTR_MASTER			(1 << 5)
+#define NH_ATTR_FLAG_FDB		(1 << 6)
+#define NH_ATTR_FLAG_BLACKHOLE		(1 << 7)
+#define NH_ATTR_FLAG_GROUPS		(1 << 8)
+/** @endcond */
+
+struct nla_policy rtnl_nh_policy[NHA_MAX+1] = {
+	[NHA_UNSPEC]		= { .type = NLA_UNSPEC },
+	[NHA_ID]		= { .type = NLA_U32 },
+	[NHA_GROUP]		= { .type = NLA_NESTED },
+	[NHA_GROUP_TYPE] 	= { .type = NLA_U16 },
+	[NHA_BLACKHOLE]  	= { .type = NLA_UNSPEC },
+	[NHA_OIF]		= { .type = NLA_U32 },
+	[NHA_MASTER]		= { .type = NLA_U32 },
+};
+
+static struct nl_cache_ops rtnl_nh_ops;
+static struct nl_object_ops nh_obj_ops;
+
+static nl_nh_group_t *rtnl_nh_grp_alloc(int len)
+{
+	nl_nh_group_t *nhg;
+
+	if (!(nhg = calloc(1, sizeof(*nhg))))
+		return NULL;
+
+	if (!(nhg->entries = calloc(len, sizeof(*nhg->entries))))
+	{
+		free (nhg);
+		return NULL;
+	}
+
+	nhg->ce_refcnt = 1;
+
+	NL_DBG(4, "Allocated new NHG object %p\n", nhg);
+
+	return nhg;
+}
+
+static void rtnl_nh_grp_free(nl_nh_group_t *nhg)
+{
+	if (!nhg)
+		return;
+
+	if (nhg->ce_refcnt > 0)
+		NL_DBG(1, "Warning: Freeing NHG object in use...\n");
```

Comment:
there is only one caller of the function, and that one already checks that the refcount is not positive. This code can never be reached.

Also, a logging about a bug is not great. Let's do `_nl_assert(nhg->ce_refcnt == 0);`

(note that the caller actually only checks that the ce_refcnt is `<= 0`. But it really can also never be negative, that would be a bug too, and the assert will be more strict than the caller).

Same for the check of `if (!nhg)`. This is a static function (called from a few lines below). You can easily see, that the condition can never be true. An `_nl_assert(nhg)` would be more appropriate.


`_nl_assert()` is usually disabled in release builds (unless the user builds with `CFLAGS=-DNL_MORE_ASSERTS=1000`, which is uncommon).

Reply:
done